include/view.h

```diff
@@ -160,6 +160,66 @@ struct xdg_toplevel_view {
 	struct wl_listener new_popup;
 };
 
+enum lab_view_criteria {
+	LAB_VIEW_CRITERIA_ALL = 0,
+	LAB_VIEW_CRITERIA_CURRENT_WORKSPACE = 1 << 0,
+	LAB_VIEW_CRITERIA_NO_ALWAYS_ON_TOP = 1 << 1,
+	LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER = 1 << 2,
+};
+
+/**
+ * for_each_view() - iterate over all views which match criteria
+ * @head: list head to iterate over
+ * @view: iterator
+ * @criteria: criteria to match against
+ */
+#define for_each_view(head, view, criteria)		\
+	for (view = view_first(head, criteria);		\
+		&view->link != (head);			\
+		view = view_next(view, criteria))
```

Comment:
Both of the view_ variants may return `NULL` based on the docs so the condition must contain a view check. I am not sure if the wrongly casted `* server->views` to a `struct view *` works correctly either. Maybe the condition could just be `view` here? Then we'd need to supply `head` as another argument to `view_next()` as well and that one could then do the check internally and return `NULL` if we reached the list head.

Reply:
I've combined `view_next()` + `view_first()` now. Getting rid of that recursion helped me see the similarity and how to deal with the first item in the list.

I think I've covered what you've said here.

The middle condition of the for loop needs to be `view && &view->link != (head)` otherwise we don't handle the case of an empty list properly. We could probably add a `wl_list_length()` check in `view_next()` instead, but I think it's simpler like this.

Reply:
> The middle condition of the for loop needs to be `view && &view->link != (head)` otherwise we don't handle the case of an empty list properly. We could probably add a `wl_list_length()` check in `view_next()` instead, but I think it's simpler like this.

I don't see why `view != NULL` wouldn't work here?

Reply:
Done